### ISSUE TITLE
Gradient of Pipek-Mezey orbital localization is off by a factor of 2

### ISSUE THREAD

**Issue by hongzhouye** [pyscf/pyscf#1984]
Hi, I found that the `get_grad` function of `lo.pipek.PipekMezey` gives gradients that seem to be off by a factor of 2 compared to numerical gradients. Is there a reason for that?

For example, these are results for water/sto-3g.
```
Exponent = 2
        PySCF     PySCFx2   Num grad 
0      -0.00000  -0.00001  -0.00001
1       0.00000   0.00000  -0.00000
2       0.00000   0.00000  -0.00000
3      -0.00000  -0.00000  -0.00000
4       0.00000   0.00000   0.00000
5       0.00000   0.00000  -0.00000
6       0.00000   0.00001   0.00001
7       0.00000   0.00001   0.00001
8      -0.00000  -0.00000  -0.00000
9      -0.00000  -0.00000  -0.00000
10     -0.14417  -0.28833  -0.28833
11     -0.01879  -0.03759  -0.03759
12      0.00000   0.00000  -0.00000
13      0.19842   0.39684   0.39684
14     -0.15528  -0.31056  -0.31056
15     -0.14417  -0.28833  -0.28833
16     -0.01879  -0.03759  -0.03759
17     -0.00000  -0.00000  -0.00000
18     -0.19842  -0.39684  -0.39684
19     -0.15528  -0.31056  -0.31056
20      0.00000   0.00000  -0.00000

Exponent = 4
        PySCF     PySCFx2   Num grad 
0      -0.00001  -0.00002  -0.00002
1       0.00000   0.00000  -0.00000
2       0.00000   0.00000  -0.00000
3      -0.00000  -0.00000  -0.00000
4       0.00000   0.00000   0.00000
5       0.00000   0.00000  -0.00000
6       0.00001   0.00003   0.00003
7       0.00001   0.00003   0.00003
8      -0.00000  -0.00000  -0.00000
9      -0.00000  -0.00000   0.00000
10     -0.28698  -0.57397  -0.57396
11     -0.03749  -0.07497  -0.07497
12      0.00000   0.00000  -0.00000
13      0.39428   0.78856   0.78855
14     -0.30901  -0.61802  -0.61802
15     -0.28698  -0.57397  -0.57396
16     -0.03749  -0.07497  -0.07497
17     -0.00000  -0.00000  -0.00000
18     -0.39428  -0.78856  -0.78855
19     -0.30901  -0.61802  -0.61802
20      0.00000   0.00000  -0.00000
```

Code for generating the results:
``` Python
import sys
import numpy as np

from pyscf import gto, scf, lo
from pyscf import lib
from pyscf.tools.dump_mat import dump_rec


if __name__ == '__main__':
    atom = '''
    O          0.00000        0.00000        0.11779
    H          0.00000        0.75545       -0.47116
    H          0.00000       -0.75545       -0.47116
    '''
    basis = 'sto-3g'

    mol = gto.M(atom=atom, basis=basis)
    mf = scf.RHF(mol).density_fit()
    mf.kernel()

    nocc = mol.nelectron // 2
    orbocc = mf.mo_coeff[:,:nocc]

    orbiao = lo.iao.iao(mol, orbocc)
    s = mf.get_ovlp()
    orbiao = lo.orth.vec_lowdin(orbiao, s)
    u0 = np.eye(orbiao.shape[1])

    for exponent in [2,4]:

        mlo = lo.pipek.PipekMezey(mol, orbiao).set(exponent=exponent)
        cost_ref = mlo.cost_function()
        g_ref, h_op_ref, h_diag_ref = mlo.gen_g_hop(u0)

        x0 = np.zeros(g_ref.size)
        step_length = 1e-3
        g_fd = np.zeros_like(g_ref)
        for i in range(g_ref.size):
            dx = np.zeros_like(x0)
            dx[i] = 1
            uf = mlo.extract_rotation(x0+dx*step_length)
            yf = mlo.cost_function(uf)
            ub = mlo.extract_rotation(x0-dx*step_length)
            yb = mlo.cost_function(ub)
            g_fd[i] = (yf - yb) / (2.*step_length)
        g_fd = -g_fd    # negative sign for maximization

        # Same results from numdifftools.Gradient
        # import numdifftools
        # def cost_func(x):
        #     u = mlo.extract_rotation(x)
        #     y = mlo.cost_function(u)
        #     return y
        # grad_func = numdifftools.Gradient(cost_func)
        # x0 = np.zeros(g_ref.size)
        # g_fd = -grad_func(x0)   # negative sign for maximization

        g_cmp = np.vstack((g_ref,g_ref*2,g_fd)).T

        print(f'Exponent = {exponent}')
        dump_rec(sys.stdout, g_cmp, label2=['PySCF','PySCFx2','Num grad'])
        print()
```

**Comment by fishjojo** [comment]
hessian is also multiplied by 1/2.

**Comment by verena-neufeld** [comment]
Hi @hongzhouye , thank you very much for posting this issue and #1990 ! Has it been resolved now by the pull request?

**Comment by hongzhouye** [comment]
Yes this issue was resolved by #1990. I am closing it now.
